corehq/apps/users/models.py

```diff
@@ -2772,3 +2773,74 @@ def can_view_web_uers(self):
 
     def can_view_roles(self):
         return False
+
+
+class UserReportingMetadataStaging(models.Model):
+    domain = models.TextField()
+    user_id = models.TextField()
+    app_id = models.TextField()
+    modified_on = models.DateTimeField(auto_now=True)
+
+    # should build_id actually be nullable?
+    build_id = models.TextField(null=True)
+
+    # The following properties are null if a user has not submitted a form since their last sync
+    xform_version = models.IntegerField(null=True)
+    form_meta = JSONField(null=True)  # This could be filtered to only the parts we need
+    received_on = models.DateTimeField(null=True)
+
+    # The following properties are null if a user has not synced since their last form submission
+    device_id = models.TextField(null=True)
+    sync_date = models.DateTimeField(null=True)
+
+    @classmethod
+    def add_submission(cls, domain, user_id, app_id, build_id, version, metadata, received_on):
+        obj, created = cls.objects.get_or_create(
+            domain=domain, user_id=user_id, app_id=app_id,
+            defaults={
+                'build_id': build_id,
+                'xform_version': version,
+                'form_meta': metadata,
+                'received_on': received_on,
+            }
+        )
+
+        if created:
+            return
+
+        save = False
+
+        if received_on - obj.received_on > settings.USER_REPORTING_METADATA_UPDATE_FREQUENCY:
```

Comment:
I think we may be able to get rid of this setting and just always update, but holding off on that for now. We can test that out by slowly reducing this on ICDS.